File: sdk/go/contract/contract.go
// Author: Tom Sapletta · https://tom.sapletta.com
// Part of the ifURI solution.
//
// Package contract: runtime guard koperty dla connectorów Go — port 1:1 kernela
// urirun_contract.gate. Connector woła EnvelopeViolation(contract, env) przed zwróceniem
// koperty; jeśli kształt rozjedzie się z kontraktem, dostaje komunikat zamiast cichego dryfu.
// Ten sam neutralny contracts.json + złote przykłady co Python/JS → walidator neutralny.
//
// Tarcie Go: JSON liczby dekodują się do float64 (int = sprawdź całkowitość); brak-vs-zero
// zachowany przez walidację map[string]interface{} (obecność klucza), nie struct.
package contract

import (
	"fmt"
	"math"
	"strconv"
	"strings"
)

func parseConst(tok string) interface{} {
	switch tok {
	case "true":
		return true
	case "false":
		return false
	}
	if i, err := strconv.Atoi(tok); err == nil {
		return i
	}
	return tok
}

func constEqual(exp, v interface{}) bool {
	if ei, ok := exp.(int); ok { // const liczbowy: JSON daje float64
		if vf, ok := v.(float64); ok {
			return float64(ei) == vf
		}
		return false
	}
	return exp == v
}

func leafOK(tok string, v interface{}) bool {
	if strings.HasPrefix(tok, "?") {
		return v == nil || leafOK(tok[1:], v)
	}
	if strings.HasPrefix(tok, "const:") {
		return constEqual(parseConst(tok[6:]), v)
	}
	if strings.HasPrefix(tok, "enum:") {
		s, ok := v.(string)
		if !ok {
			return false
		}
		for _, a := range strings.Split(tok[5:], "|") {
			if a == s {
				return true
			}
		}
		return false
	}
	switch tok {
	case "str":
		_, ok := v.(string)
		return ok
	case "int":
		f, ok := v.(float64)
		return ok && f == math.Trunc(f)
	case "num":
		_, ok := v.(float64)
		return ok
	case "bool":
		_, ok := v.(bool)
		return ok
	case "obj":
		_, ok := v.(map[string]interface{})
		return ok
	case "list":
		_, ok := v.([]interface{})
		return ok
	case "any":
		return true
	}
	return false
}

// Check zwraca błąd z umiejscowionym komunikatem, gdy value nie spełnia schema (jak kernel `check`).
func Check(schema, value interface{}, where string) error {
	if m, ok := schema.(map[string]interface{}); ok {
		if alts, has := m["oneOf"]; has {
			var errs []string
			for i, alt := range alts.([]interface{}) {
				if e := Check(alt, value, fmt.Sprintf("%s|oneOf[%d]", where, i)); e == nil {
					return nil
				} else {
					errs = append(errs, e.Error())
				}
			}
			return fmt.Errorf("%s: matched none of oneOf -> %v", where, errs)
		}
		vm, ok := value.(map[string]interface{})
		if !ok {
			return fmt.Errorf("%s: expected object", where)
		}
		for key, sub := range m {
			subStr, isStr := sub.(string)
			if isStr && strings.HasPrefix(subStr, "?") {
				if _, present := vm[key]; !present {
					continue
				}
			}
			val, present := vm[key]
			if !present {
				return fmt.Errorf("%s: missing required key %q", where, key)
			}
			if e := Check(sub, val, fmt.Sprintf("%s.%s", where, key)); e != nil {
				return e
			}
		}
		return nil
	}
	if tok, ok := schema.(string); ok {
		if !leafOK(tok, value) {
			return fmt.Errorf("%s: %v does not satisfy %s", where, value, tok)
		}
		return nil
	}
	if arr, ok := schema.([]interface{}); ok && len(arr) > 0 {
		va, ok := value.([]interface{})
		if !ok {
			return fmt.Errorf("%s: expected list", where)
		}
		for i, it := range va {
			if e := Check(arr[0], it, fmt.Sprintf("%s[%d]", where, i)); e != nil {
				return e
			}
		}
		return nil
	}
	return nil
}

// EnvelopeViolation zwraca komunikat naruszenia albo "" (jak kernel `envelope_violation`).
func EnvelopeViolation(c map[string]interface{}, env map[string]interface{}) string {
	if ok, _ := env["ok"].(bool); ok {
		if out, ok := c["out"].(map[string]interface{}); ok && len(out) > 0 {
			if e := Check(out, env, "out"); e != nil {
				return e.Error()
			}
		}
		return ""
	}
	var cls interface{}
	if rem, ok := env["remediation"].(map[string]interface{}); ok {
		cls = rem["class"]
	}
	if cls == nil {
		if errm, ok := env["error"].(map[string]interface{}); ok {
			cls = errm["remediationClass"]
		}
	}
	if errors, ok := c["errors"].([]interface{}); ok && len(errors) > 0 && cls != nil {
		for _, e := range errors {
			if e == cls {
				return ""
			}
		}
		return fmt.Sprintf("error class %v not in declared %v", cls, errors)
	}
	return ""
}

File: sdk/js/contract.mjs
#!/usr/bin/env node
// Author: Tom Sapletta · https://tom.sapletta.com
// Part of the ifURI solution.
//
// Runtime guard koperty dla connectorów JS — port 1:1 kernela urirun_contract.gate.
// Connector woła `envelopeViolation(contract, env)` PRZED zwróceniem koperty: jeśli kształt
// rozjedzie się z kontraktem, dostaje komunikat (nie cichy dryf). Ten sam contracts.json,
// te same złote przykłady co Python/Go → walidator jest językowo-neutralny.
//
// CLI (do testów parytetu): node contract.mjs <contracts.json> <route>   # koperta z stdin
//   wypisuje "OK" (exit 0) albo "VIOLATION <msg>" (exit 1).

const parseConst = (tok) =>
  tok === "true" ? true : tok === "false" ? false
  : /^-?\d+$/.test(tok) ? parseInt(tok, 10) : tok;

function leafOk(token, value) {
  if (token.startsWith("?")) return value === null || value === undefined || leafOk(token.slice(1), value);
  if (token.startsWith("const:")) return value === parseConst(token.slice(6));
  if (token.startsWith("enum:")) return token.slice(5).split("|").includes(value);
  switch (token) {
    case "str": return typeof value === "string";
    case "int": return typeof value === "number" && Number.isInteger(value);
    case "num": return typeof value === "number";
    case "bool": return typeof value === "boolean";
    case "obj": return value !== null && typeof value === "object" && !Array.isArray(value);
    case "list": return Array.isArray(value);
    case "any": return true;
  }
  return false;
}

// Rzuca Error z umiejscowionym komunikatem (jak kernel `check`).
export function check(schema, value, where) {
  if (schema && typeof schema === "object" && !Array.isArray(schema)) {
    if ("oneOf" in schema) {
      const errs = [];
      for (let i = 0; i < schema.oneOf.length; i++) {
        try { check(schema.oneOf[i], value, `${where}|oneOf[${i}]`); return; }
        catch (e) { errs.push(e.message); }
      }
      throw new Error(`${where}: matched none of oneOf -> ${errs}`);
    }
    if (value === null || typeof value !== "object" || Array.isArray(value))
      throw new Error(`${where}: expected object`);
    for (const [key, spec] of Object.entries(schema)) {
      if (!(key in value)) {
        if (typeof spec === "string" && spec.startsWith("?")) continue;
        throw new Error(`${where}: missing required key '${key}'`);
      }
      check(spec, value[key], `${where}.${key}`);
    }
    return;
  }
  if (Array.isArray(schema)) {
    if (!Array.isArray(value)) throw new Error(`${where}: expected list`);
    if (schema.length) value.forEach((it, i) => check(schema[0], it, `${where}[${i}]`));
    return;
  }
  if (!leafOk(schema, value))
    throw new Error(`${where}: ${JSON.stringify(value)} does not satisfy ${schema}`);
}

// Zwraca komunikat naruszenia albo null (jak kernel `envelope_violation`).
export function envelopeViolation(contract, env) {
  try {
    if (env && env.ok) {
      if (contract.out && Object.keys(contract.out).length) check(contract.out, env, "out");
      return null;
    }
    let cls = null;
    const rem = env && env.remediation;
    if (rem && typeof rem === "object") cls = rem.class;
    if (cls == null) {
      const err = env && env.error;
      if (err && typeof err === "object") cls = err.remediationClass;
    }
    const errors = contract.errors || [];
    if (errors.length && cls != null && !errors.includes(cls))
      return `error class '${cls}' not in declared ${JSON.stringify(errors)}`;
  } catch (e) { return e.message; }
  return null;
}

// Neutralny contracts.json → {route: contract} (te same pola co kernel).
export function loadContracts(doc) {
  const out = {};
  for (const [route, c] of Object.entries(doc.contracts || {})) {
    out[route] = {
      version: c.version || "v1", effect: c.effect || "query",
      reversible: !!c.reversible, inverseRoute: c.inverseRoute || "",
      inp: c.inp || {}, out: c.out || {}, errors: c.errors || [], examples: c.examples || [],
    };
  }
  return out;
}

if (import.meta.url === `file://${process.argv[1]}`) {
  const fs = await import("node:fs");
  const [, , contractsPath, route] = process.argv;
  const contracts = loadContracts(JSON.parse(fs.readFileSync(contractsPath, "utf8")));
  let input = "";
  for await (const chunk of process.stdin) input += chunk;
  const violation = envelopeViolation(contracts[route], JSON.parse(input));
  if (violation) { console.log("VIOLATION " + violation); process.exit(1); }
  console.log("OK");
}
